### ISSUE TITLE
[Feature] Computing partial trace outputs for statevectors and density matrices

### ISSUE THREAD

**Issue by rajaiitp** [pasqal-io/qadence#367]
It would nice to have basic density matrix related functionalities that are part of basic quantum information theory computations (partial trace, entropy, fidelity, purity..)

**Comment by Roland-djee** [comment]
@rajaiitp Could you please be more specific about the mathematical formulation of these ? Thanks.

**Comment by rajaiitp** [comment]
Partial trace of qubits marked by index i,j
Partial trace_{ij}(rho_ijkl) =  Trace_{ij}(rho_ijkl) =Sum_{i=(0,1),j=(0,1)} rho_ijkl 

Von-Neuman Entropy of density matrix 
Entropy = - trace( rho log rho)
trace over all qubit indices.

Fidelity between 2 density matrices rho, sigma
Fidelity(rho,sigma) = (tr(sqrt(sqrt(rho)sigma sqrt(simga)))^2

here sqrt is defined as diagonalize, sqrt and undiagonalize

n_th purity of a density matrix rho,
Purity(n,rho) = tr(rho^n)

**Comment by Roland-djee** [comment]
Hey @rajaiitp could you please make sure the formulae above are correct ? Thanks.

**Comment by Roland-djee** [comment]
In my opinion, this could go with the introduction of a `Result` type.

**Comment by rajaiitp** [comment]
@Roland-djee yes the formulae above are correct, just edited some brackets, didnt know we had support for latex on here. I dont know what a Result type is

**Comment by mlahariya** [comment]
@rajaiitp Is this still needed in qadence? 

We could resolve this in qadence 2 using runtimes.

**Comment by rajaiitp** [comment]
Rarely used. Its just a property of the state that sometimes people care to look into. (probably used once or twice so far at best)
